File: src/main/java/infinitespire/cards/black/FinalStrike.java
package infinitespire.cards.black;

import com.badlogic.gdx.math.MathUtils;
import com.megacrit.cardcrawl.actions.AbstractGameAction.AttackEffect;
import com.megacrit.cardcrawl.actions.common.DamageAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.cards.red.PerfectedStrike;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.powers.AbstractPower;
import infinitespire.abstracts.BlackCard;

public class FinalStrike extends BlackCard {

	public static final String ID = "FinalStrike";
	private static final String NAME = "Final Strike";
	private static final String IMG = "img/infinitespire/cards/finalstrike.png";
	private static final int COST = 2;
											//Basically a better perfected strike
	private static final String DESCRIPTION = "Deal !D! damage. Deals an additional !M! damage for ALL of your cards containing Strike.";
	private static final CardType TYPE = CardType.ATTACK;
	private static final CardTarget TARGET = CardTarget.ENEMY;
	
	public FinalStrike() {
		super(ID, NAME, IMG, COST, DESCRIPTION, TYPE, TARGET);
		this.baseDamage = 10;
		this.baseMagicNumber = 5;
		this.magicNumber = 5;
		this.tags.add(CardTags.STRIKE);
	}
	
	@Override
	public AbstractCard makeCopy() {
		return new FinalStrike();
	}
	
	public static boolean isStrike(AbstractCard card) {
		boolean retVal = PerfectedStrike.isStrike(card);
		
		if(card.cardID.toLowerCase().contains("strike")) {
			return true;
		}
		
		return retVal;
	}
	
	

	@Override
	public void applyPowers() {
		float tmp = this.baseDamage;
		tmp += this.magicNumber * countCards();
		for(AbstractPower p : AbstractDungeon.player.powers) {
			tmp = p.atDamageGive(tmp, this.damageTypeForTurn);
		}
		for(AbstractPower p : AbstractDungeon.player.powers) {
			tmp = p.atDamageFinalGive(tmp, this.damageTypeForTurn);
		}
		
		if(tmp < 0.0f) {
			tmp = 0.0f;
		}
		
		if(this.baseDamage != (int)tmp) {
			this.isDamageModified = true;
		}
		
		this.damage = MathUtils.floor(tmp);
	}
	
	

	@Override
	public void calculateCardDamage(AbstractMonster m) {
		float tmp = this.baseDamage;
		tmp += this.magicNumber * countCards();
		for(AbstractPower p : AbstractDungeon.player.powers) {
			tmp = p.atDamageGive(tmp, this.damageTypeForTurn);
		}
		for(AbstractPower p : AbstractDungeon.player.powers) {
			tmp = p.atDamageFinalGive(tmp, this.damageTypeForTurn);
		}
		if(m != null) {
			for(AbstractPower p : m.powers) {
				tmp = p.atDamageReceive(tmp, this.damageTypeForTurn);
			}
			for(AbstractPower p : m.powers) {
				tmp = p.atDamageFinalReceive(tmp, this.damageTypeForTurn);
			}
		}
		if(tmp < 0.0f) {
			tmp = 0.0f;
		}
		
		if(this.baseDamage != (int)tmp) {
			this.isDamageModified = true;
		}
		
		this.damage = MathUtils.floor(tmp);
	}

	private static int countCards() {
		int count = 0;
		for(AbstractCard card : AbstractDungeon.player.hand.group)
			if(isStrike(card) || card.tags.contains(CardTags.STRIKE)) count++;
		
		for(AbstractCard card : AbstractDungeon.player.drawPile.group) 
			if(isStrike(card) || card.tags.contains(CardTags.STRIKE)) count++;
		
		for(AbstractCard card : AbstractDungeon.player.discardPile.group)
			if(isStrike(card) || card.tags.contains(CardTags.STRIKE)) count++;
		
		return count;
	}

	@Override
	public void upgrade() {
		if(!this.upgraded) {
			this.upgradeName();
			this.upgradeDamage(5);
		}
	}

	@Override
	public void useWithEffect(AbstractPlayer p, AbstractMonster m) {
		AbstractDungeon.actionManager.addToBottom(new DamageAction(m, new DamageInfo(p, this.damage, this.damageTypeForTurn), AttackEffect.SLASH_DIAGONAL));
	}

}

File: src/main/kotlin/com/blanktheevil/infinitespire/cards/black/FinalStrike.kt
package com.blanktheevil.infinitespire.cards.black

import com.blanktheevil.infinitespire.extensions.makeID
import com.blanktheevil.infinitespire.extensions.player
import com.blanktheevil.infinitespire.utils.CardBuilder
import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.actions.common.DamageAction
import com.megacrit.cardcrawl.actions.utility.WaitAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.cards.red.PerfectedStrike
import com.megacrit.cardcrawl.monsters.AbstractMonster
import kotlin.math.floor
import kotlin.streams.toList

@Suppress("unused")
class FinalStrike : BlackCard(BUILDER) {
  companion object {
    val ID = "FinalStrike".makeID()
    private const val DAMAGE = 10
    private const val MAGIC = 5
    private const val UPG_DAMAGE = 5
    private val BUILDER = CardBuilder(id = ID)
      .img("finalstrike.png")
      .attack()
      .enemy()
      .cost(2)
      .use { p, m ->
        addToBot(
          WaitAction(0.8f)
        )
        addToBot(
          DamageAction(
            m,
            DamageInfo(p, damage, damageTypeForTurn),
            AbstractGameAction.AttackEffect.NONE
          )
        )
      }
      .init {
        baseDamage = DAMAGE
        baseMagicNumber = MAGIC
        magicNumber = MAGIC
        this.tags.add(CardTags.STRIKE)
      }
      .upgr {
        upgradeDamage(UPG_DAMAGE)
      }

    fun isStrike(card: AbstractCard): Boolean {
      return if (PerfectedStrike.isStrike(card) || card.tags.contains(CardTags.STRIKE)) {
        true
      } else card.cardID.toLowerCase().contains("strike")
    }
  }

  private fun getStrikeDamage(): Float {
    val groups = mutableListOf<CardGroup>()
    var count = 0

    with(player) {
      groups.add(hand)
      groups.add(drawPile)
      groups.add(discardPile)
    }

    groups.forEach { cGroup ->
      count += cGroup.group.stream()
        .filter { isStrike(it) }
        .toList().size
    }

    return count.times(magicNumber.toFloat())
  }

  override fun applyPowers() {
    var tmp: Float = this.baseDamage.toFloat()
    tmp += getStrikeDamage()
    with(player) {
      powers.forEach { it.atDamageGive(tmp, damageTypeForTurn) }
      powers.forEach { it.atDamageFinalGive(tmp, damageTypeForTurn) }
    }

    tmp = if (tmp < 0f) 0f else tmp
    isDamageModified = baseDamage != tmp.toInt()
    damage = floor(tmp).toInt()
  }

  override fun calculateCardDamage(mo: AbstractMonster?) {
    var tmp: Float = this.baseDamage.toFloat()
    tmp += getStrikeDamage()
    with(player) {
      powers.forEach { it.atDamageGive(tmp, damageTypeForTurn) }
      powers.forEach { it.atDamageFinalGive(tmp, damageTypeForTurn) }
    }

    if (mo != null) {
      with(mo) {
        powers.forEach { it.atDamageReceive(tmp, damageTypeForTurn) }
        powers.forEach { it.atDamageFinalReceive(tmp, damageTypeForTurn) }
      }
    }

    tmp = if (tmp < 0f) 0f else tmp
    isDamageModified = baseDamage != tmp.toInt()
    damage = floor(tmp).toInt()
  }
}